File: Python/ChatGPT4/NBodySimulationPar.py
import numpy as np
import matplotlib.pyplot as plt
from concurrent.futures import ProcessPoolExecutor, as_completed

# Constants
G = 6.67430e-11  # Gravitational constant (m^3 kg^-1 s^-2)
mass_central = 1.989e30  # Mass of the central body (kg), e.g., mass of the Sun
time_step = 1e5  # Time step (s)
num_steps = 1000  # Number of steps to simulate
num_bodies = 1000000  # Number of small bodies

# Initialize positions and velocities of bodies
def initialize_bodies(num_bodies, mass_central):
    # Place small bodies in circular orbits around the central body
    # Assume all bodies are at the same distance from the central body, in a circular orbit
    radius = 1e11  # Distance from the central body (m), roughly 1 AU
    angles = np.linspace(0, 2 * np.pi, num_bodies, endpoint=False)
    x_positions = radius * np.cos(angles)
    y_positions = radius * np.sin(angles)
    z_positions = np.zeros(num_bodies)  # All bodies in the same plane
    
    # Velocities based on circular orbits: v = sqrt(G * M / r)
    velocities = np.sqrt(G * mass_central / radius)
    vx_positions = -velocities * np.sin(angles)
    vy_positions = velocities * np.cos(angles)
    vz_positions = np.zeros(num_bodies)  # No motion in the z-direction
    
    # Combine positions and velocities
    positions = np.array([x_positions, y_positions, z_positions]).T
    velocities = np.array([vx_positions, vy_positions, vz_positions]).T
    
    # The mass of the small bodies (arbitrary small mass for demonstration purposes)
    masses = np.full(num_bodies, 1e12)  # Small mass of each body
    
    return positions, velocities, masses

# Compute the total energy of the system (kinetic + potential)
def compute_energy(positions, velocities, masses, mass_central):
    kinetic_energy = 0.5 * np.sum(masses * np.linalg.norm(velocities, axis=1)**2)
    potential_energy = 0.0
    # Gravitational potential energy between each pair of bodies (including central body)
    for i in range(len(masses)):
        # Potential energy with the central body
        r = np.linalg.norm(positions[i])
        potential_energy -= G * mass_central * masses[i] / r
        # Inter-body potential energy (optional, but usually not significant in large N-body simulations)
        for j in range(i + 1, len(masses)):
            r_ij = np.linalg.norm(positions[i] - positions[j])
            potential_energy -= G * masses[i] * masses[j] / r_ij
    
    total_energy = kinetic_energy + potential_energy
    return total_energy

# Compute the acceleration for a specific body due to the gravitational force
def compute_acceleration(i, positions, masses, mass_central):
    num_bodies = len(masses)
    acceleration = np.zeros(3)
    
    # Gravitational acceleration from the central body
    r = np.linalg.norm(positions[i])
    accel_central = -G * mass_central * positions[i] / r**3
    acceleration += accel_central
    
    # Gravitational accelerations due to other bodies
    for j in range(num_bodies):
        if i != j:
            r_ij = positions[i] - positions[j]
            r_ij_norm = np.linalg.norm(r_ij)
            accel_interbody = G * masses[j] * r_ij / r_ij_norm**3
            acceleration += accel_interbody
    
    return i, acceleration

# Perform the simulation using the kick-step method with parallelized force computation
def run_simulation(positions, velocities, masses, mass_central, num_steps, time_step):
    # Store energy before the simulation
    initial_energy = compute_energy(positions, velocities, masses, mass_central)
    
    # Perform the time integration
    with ProcessPoolExecutor() as executor:
        for step in range(num_steps):
            # Parallelize the acceleration computation
            futures = []
            for i in range(len(masses)):
                futures.append(executor.submit(compute_acceleration, i, positions, masses, mass_central))
            
            # Collect accelerations after all tasks are completed
            accelerations = np.zeros_like(positions)
            for future in as_completed(futures):
                i, acceleration = future.result()
                accelerations[i] = acceleration
            
            # First kick: Update velocities (half step)
            velocities += 0.5 * accelerations * time_step
            
            # Update positions: full step
            positions += velocities * time_step
            
            # Recompute accelerations at new positions
            futures = []
            for i in range(len(masses)):
                futures.append(executor.submit(compute_acceleration, i, positions, masses, mass_central))
            
            # Collect accelerations after all tasks are completed
            accelerations = np.zeros_like(positions)
            for future in as_completed(futures):
                i, acceleration = future.result()
                accelerations[i] = acceleration
            
            # Second kick: Update velocities (half step)
            velocities += 0.5 * accelerations * time_step
    
    # Compute energy after the simulation
    final_energy = compute_energy(positions, velocities, masses, mass_central)
    
    return initial_energy, final_energy, positions, velocities

# Main function to initialize and run the simulation
def main():
    positions, velocities, masses = initialize_bodies(num_bodies, mass_central)
    
    # Perform the simulation
    initial_energy, final_energy, positions, velocities = run_simulation(positions, velocities, masses, mass_central, num_steps, time_step)
    
    # Display energy before and after the simulation
    print(f"Initial Energy: {initial_energy:.2e} J")
    print(f"Final Energy: {final_energy:.2e} J")
    print(f"Energy difference: {final_energy - initial_energy:.2e} J")
    
    # Optional: Plot the positions of the bodies
    plt.figure(figsize=(8, 8))
    plt.scatter(positions[:, 0], positions[:, 1], s=0.1, color='blue', alpha=0.5)
    plt.title("Positions of Bodies after Simulation")
    plt.xlabel("X (m)")
    plt.ylabel("Y (m)")
    plt.axis('equal')
    plt.show()

if __name__ == "__main__":
    main()

File: Java/Gemini2/NBodySimulationPar.java
import java.util.Random;
import java.util.Arrays;
import java.util.stream.IntStream;

public class NBodySimulationPar {

    private static final double GRAVITATIONAL_CONSTANT = 6.674e-11;
    private static final double TIME_STEP = 0.001; // Adjust as needed
    private static final int NUM_STEPS = 1000;

    private Body[] bodies;

    public NBodySimulationPar(int numBodies) {
        bodies = new Body[numBodies + 1];
        initializeSystem();
    }

    private void initializeSystem() {
        Random random = new Random(0); // Seed for reproducibility

        // Initialize central body (Sun-like)
        bodies[0] = new Body(1.989e30, 0, 0, 0, 0, 0, 0); // Mass, x, y, z, vx, vy, vz

        // Initialize smaller bodies (planet-like) in circular orbits
        for (int i = 1; i < bodies.length; i++) {
            double mass = 1e25 + random.nextDouble() * 1e26; // Varying mass
            double radius = 1e11 + random.nextDouble() * 1e12; // Varying orbital radius
            double speed = Math.sqrt(GRAVITATIONAL_CONSTANT * bodies[0].mass / radius);

            double angle = random.nextDouble() * 2 * Math.PI;
            double x = radius * Math.cos(angle);
            double y = radius * Math.sin(angle);
            double z = 0; // Orbit in x-y plane for simplicity

            double vx = -speed * Math.sin(angle);
            double vy = speed * Math.cos(angle);
            double vz = 0;

            bodies[i] = new Body(mass, x, y, z, vx, vy, vz);
        }
    }

    public void simulate() {
        double initialEnergy = calculateTotalEnergy();
        System.out.println("Initial Energy: " + initialEnergy);

        for (int step = 0; step < NUM_STEPS; step++) {
            // Kick step (parallelized)
            IntStream.range(0, bodies.length).parallel().forEach(i -> {
                double fx = 0, fy = 0, fz = 0;
                for (int j = 0; j < bodies.length; j++) {
                    if (i == j) continue;

                    double dx = bodies[j].x - bodies[i].x;
                    double dy = bodies[j].y - bodies[i].y;
                    double dz = bodies[j].z - bodies[i].z;
                    double distSq = dx * dx + dy * dy + dz * dz;
                    double forceMagnitude = GRAVITATIONAL_CONSTANT * bodies[i].mass * bodies[j].mass / distSq;

                    fx += forceMagnitude * dx / Math.sqrt(distSq);
                    fy += forceMagnitude * dy / Math.sqrt(distSq);
                    fz += forceMagnitude * dz / Math.sqrt(distSq);
                }

                bodies[i].vx += fx / bodies[i].mass * TIME_STEP;
                bodies[i].vy += fy / bodies[i].mass * TIME_STEP;
                bodies[i].vz += fz / bodies[i].mass * TIME_STEP;
            });

            // Drift step (parallelized - simpler, can also be parallelized)
            IntStream.range(0, bodies.length).parallel().forEach(i -> {
                bodies[i].x += bodies[i].vx * TIME_STEP;
                bodies[i].y += bodies[i].vy * TIME_STEP;
                bodies[i].z += bodies[i].vz * TIME_STEP;
            });
        }

        double finalEnergy = calculateTotalEnergy();
        System.out.println("Final Energy: " + finalEnergy);
        System.out.println("Energy Change: " + (finalEnergy - initialEnergy));
    }


    private double calculateTotalEnergy() {
        double kineticEnergy = Arrays.stream(bodies).parallel()
                .mapToDouble(body -> 0.5 * body.mass * (body.vx * body.vx + body.vy * body.vy + body.vz * body.vz))
                .sum();


        double potentialEnergy = IntStream.range(0, bodies.length).parallel()
                .mapToDouble(i -> {
                    double pe = 0;
                    for (int j = i + 1; j < bodies.length; j++) {
                        double dx = bodies[j].x - bodies[i].x;
                        double dy = bodies[j].y - bodies[i].y;
                        double dz = bodies[j].z - bodies[i].z;
                        double dist = Math.sqrt(dx * dx + dy * dy + dz * dz);
                        pe -= GRAVITATIONAL_CONSTANT * bodies[i].mass * bodies[j].mass / dist;
                    }
                    return pe;
                }).sum();

        return kineticEnergy + potentialEnergy;
    }

    private static class Body {
        double mass, x, y, z, vx, vy, vz;

        public Body(double mass, double x, double y, double z, double vx, double vy, double vz) {
            this.mass = mass;
            this.x = x;
            this.y = y;
            this.z = z;
            this.vx = vx;
            this.vy = vy;
            this.vz = vz;
        }
    }

    public static void main(String[] args) {
        int numSmallBodies = 100000;  // One million small bodies
        NBodySimulationPar simulation = new NBodySimulationPar(numSmallBodies);
        simulation.simulate();
    }
}
